<a href="https://colab.research.google.com/github/srishti73/Machine-Learning/blob/main/Multiple_Linear_Regression.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

5 steps in multiple regression
1)All in
2) Backward elimination
3)forward selection
4)bidirectional elimination
5)score comparison
2,3,4 are called stepwise regression

In [5]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np

In [6]:
dataset = pd.read_csv('50_Startups.csv')
x=dataset.iloc[:,:-1].values
y=dataset.iloc[:,-1].values

In [7]:
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder
ct=ColumnTransformer(transformers=[('encoder',OneHotEncoder(),[3])], remainder='passthrough')
x=np.array(ct.fit_transform(x))
#[3] is the column having categorical value

In [8]:
from sklearn.model_selection import train_test_split
x_train, x_test, y_train,y_test=train_test_split(x,y,test_size=0.2,random_state=0)

In [10]:
#Train mlr model
from sklearn.linear_model import LinearRegression 
regressor= LinearRegression() #builds mlr model
regressor.fit(x_train, y_train) #trains mlr on training set



LinearRegression()

In [14]:
y_pred= regressor.predict(x_test) #vector of predicted profit in the test set
np.set_printoptions(precision=2)
 #will display any numerical value with only 2 decimal after comma
print(np.concatenate((y_pred.reshape(len(y_pred),1), y_test.reshape(len(y_test),1)), 1))
#gives 2 vectors, 1st being predicted profit and 2nd being tested profit


[[103015.2  103282.38]
 [132582.28 144259.4 ]
 [132447.74 146121.95]
 [ 71976.1   77798.83]
 [178537.48 191050.39]
 [116161.24 105008.31]
 [ 67851.69  81229.06]
 [ 98791.73  97483.56]
 [113969.44 110352.25]
 [167921.07 166187.94]]


Making a single prediction (for example the profit of a startup with R&D Spend = 160000, Administration Spend = 130000, Marketing Spend = 300000 and State = 'California')

In [15]:
print(regressor.predict([[1,0,0,160000,130000,300000]]))
# no bracket= scalar, 1 bracket [] = 1d array , 2 brackets [[]]= 2d array

[181566.92]


Getting the final linear regression equation with the values of the coefficients

In [16]:
print(regressor.coef_)
print(regressor.intercept_)

[ 8.66e+01 -8.73e+02  7.86e+02  7.73e-01  3.29e-02  3.66e-02]
42467.52924853204


Therefore, the equation of our multiple linear regression model is:

Profit=86.6×Dummy State 1−873×Dummy State 2+786×Dummy State 3+0.773×R&D Spend+0.0329×Administration+0.0366×Marketing Spend+42467.53